runtime/master/src/main/java/org/apache/nemo/runtime/master/scheduler/BatchScheduler.java

```diff
@@ -447,6 +424,60 @@ private void doSchedule() {
     return tasks;
   }
 
+  ////////////////////////////////////////////////////////////////////// Task cloning methods.
+
+  /**
+   * @return true if the number of clones is modified.
+   *         false otherwise.
+   */
+  private boolean modifyStageNumClone(final Stage stage, final ClonedSchedulingProperty.CloneConf cloneConf) {
+    final double fractionToWaitFor = cloneConf.getFractionToWaitFor();
+    final Object[] completedTaskTimes = planStateManager.getCompletedTaskTimeListMs(stage.getId()).toArray();
+
+    // Only after the fraction of the tasks are done...
+    // Delayed cloning (aggressive)
+    if (completedTaskTimes.length > 0
+      && completedTaskTimes.length >= Math.round(stage.getTaskIndices().size() * fractionToWaitFor)) {
+      // Only if the running task is considered a 'straggler'....
+      Arrays.sort(completedTaskTimes);
+      final long medianTime = (long) completedTaskTimes[completedTaskTimes.length / 2];
+      final double medianTimeMultiplier = cloneConf.getMedianTimeMultiplier();
+      final Map<String, Long> executingTaskToTime = planStateManager.getExecutingTaskToRunningTimeMs(stage.getId());
+
+      return modifyStageNumCloneUsingMedianTime(
+        stage.getId(), completedTaskTimes.length, medianTime, medianTimeMultiplier, executingTaskToTime);
+    } else {
+      return false;
+    }
+  }
+
+  /**
+   * @return true if the number of clones is modified.
+   *         false otherwise.
+   */
+  private boolean modifyStageNumCloneUsingMedianTime(final String stageId,
```

Comment:
Do we need to separate this part from a `modifyStageNumClone`?

The both function has too similar function name and exactly same function description..

IMHO, it's better to make them as separate functions when we can clearly define them.. in other words, when we can give them clear names and descriptions.

Reply:
We probably need this to avoid the "Cognitive Complexity" issue.

Link to the SonarCloud issue.
https://sonarcloud.io/project/issues?id=org.apache.nemo%3Anemo-project&issues=AWkltI_5gtFwgfOLu2KZ&open=AWkltI_5gtFwgfOLu2KZ

Reply:
I've changed some of the names to make them more descriptive.